src/minimum_tracer/minimum_tracer.cpp

```diff
@@ -0,0 +1,3214 @@
+        &Hessian,
+    double error,
+    double const_multiplier,
+    int maxiter)
+{
+  // Checks if guess is close enough
+  if (L2NormVector(df(guess)) < error) return guess;
+
+  // If not, performs gradient descent until minima with a maximum of
+  // "maxiter" iterations
+  int dim                       = guess.size(); // Guess dimension
+  int i                         = 0;            // Counter
+  std::vector<double> new_guess = guess;        // First guess
+  std::vector<double> gradient;                 // Gradient
+  std::vector<std::vector<double>> Hess;        // HessianNumerical gradient
+
+  for (i = 0; (i < maxiter) && (L2NormVector(df(new_guess)) > error); i++)
+  {
+    // std::cout << i << "\t" << L2NormVector(df(new_guess)) << "\n";
+
+    //  Update grad and HessianNumerical :
+    gradient = df(new_guess);
+    Hess     = Hessian(new_guess);
+
+    // Convert into a Eigen3 // Probably there is a better way
+    Eigen::MatrixXd HessMatrix(dim, dim);
+    for (int m = 0; m < dim; m++)
+    {
+      // EigenGradient(m) = 0.1;
+      for (int n = 0; n < dim; n++)
+      {
+        HessMatrix(m, n) = Hess[m][n];
+      }
+      HessMatrix(m, m) += 1e-3;
+    }
+
+    if (HessMatrix.determinant() != 0) // If HessianNumerical is
+    // invertible them do the gradient descent, if not use only the
+    // diagonal elements.
+    {
+      // Convert gradient into vector
+      Eigen::VectorXd EigenGradient =
+          Eigen::Map<Eigen::VectorXd, Eigen::Unaligned>(gradient.data(),
+                                                        gradient.size());
+
+      // Calculates the n - th step : x(n + 1) = x(n) + alpha(n)
+      Eigen::VectorXd delta =
+          HessMatrix.colPivHouseholderQr().solve(-1 * EigenGradient);
+
+      // HessMatrix.colPivHouseholderQr().solve(EigenGradient);
+      // alpha = -const_multiplier * HessMatrix.inverse() * gradient;
+      for (int j = 0; j < dim; j++)
+      {
+        new_guess[j] += delta(j); // Updates guess
+      }
+    }
+    else // If is not invertible use just the gradient
+    {
+      for (int j = 0; j < dim; j++)
+      {
+        new_guess[j] -= const_multiplier * gradient[j]; // Updates guess
+      }
+    }
+  }
+  return (new_guess);
+}
+
+double MinimumTracer::SmallestEigenvalue(
+    const std::vector<double> &point,
+    const std::function<std::vector<std::vector<double>>(std::vector<double>)>
+        &Hessian)
+{
+  std::size_t dim                                  = point.size();
+  std::vector<std::vector<double>> current_hessian = Hessian(point);
+
+  Eigen::MatrixXcd mat(dim, dim);
+  for (std::size_t i = 0; i < dim; i++)
+  {
+    mat.col(i) = Eigen::Map<Eigen::VectorXd>(current_hessian[i].data(), dim);
+  }
+
+  // Calculate smallest eigenvalue
+  double current_min = 1e100;
+  for (auto element : mat.eigenvalues())
+    current_min = std::min(element.real(), current_min);
+  return current_min + 1e-7;
```

Comment:
Why that extra `1e-7`?

Reply:
If the smallest eigenvalue gets to zero it can become slightly negative due to numerical errors. This prevents that.

Reply:
Alright, this sounds good. Can you maybe add this as a comment to the code? Otherwise the next person looking at this will not know this

Reply:
Done.